File: Java/Sample_Source/Images/DrawSeparations/src/com/datalogics/PDFL/Samples/DrawSeparations.java
package com.datalogics.PDFL.Samples;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

import javax.imageio.ImageIO;

import com.datalogics.PDFL.Document;
import com.datalogics.PDFL.DrawFlags;
import com.datalogics.PDFL.DrawParams;
import com.datalogics.PDFL.Ink;
import com.datalogics.PDFL.Library;
import com.datalogics.PDFL.Matrix;
import com.datalogics.PDFL.Page;
import com.datalogics.PDFL.Rect;
import com.datalogics.PDFL.SeparationColorSpace;

public class DrawSeparations {
	 
	/*
	 * This sample demonstrates drawing a list of grayscale separations from a PDF file.
	 *
	 * For more detail see the description of the DrawSeparations sample program on our Developer's site, 
	 * http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/java-sample-programs/manipulating-graphics-and-separating-colors-for-images
	 * 
	/**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        System.out.println("DrawSeparations Sample:");
        
        System.setProperty("java.awt.headless", "true");

        Library lib = new Library();

        try {
            String sInput = "../../Resources/Sample_Input/ducky.pdf";
            String sOutput = "DrawSeparations-out-";
            if (args.length > 0)
                sInput = args[0];
            if (args.length > 1)
                sOutput = args[1];
            System.out.println("Initialized the library.");
            System.out.println("PDF Document to Images Sample:");
            System.out.println("Input file: " + sInput + ", will write to " + sOutput);

            
            Document doc = new Document(sInput);
            Page pg = doc.getPage(0);
            
            // Get all inks tha are present on the page
            List<Ink> inks = pg.listInks();
            List<SeparationColorSpace> colorants = new ArrayList<SeparationColorSpace>();
            
            // Here we decide, which inks should be drawn
            for (Ink ink : inks) {
            	colorants.add(new SeparationColorSpace(pg, ink));
            }
            
            double width = pg.getMediaBox().getRight() - pg.getMediaBox().getLeft();
            double height = pg.getMediaBox().getTop() - pg.getMediaBox().getBottom();
            

            // Must invert the page to get from PDF with origin at lower left,
            // to a bitmap with the origin at upper right.
            Matrix matrix = new Matrix().scale(1, -1).translate(0, -height);
            
            DrawParams params = new DrawParams();
            params.setMatrix(matrix);
            params.setDestRect(new Rect(0, 0, width, height));
            params.setFlags(EnumSet.of(DrawFlags.DO_LAZY_ERASE, DrawFlags.USE_ANNOT_FACES));
            
            // Acquiring list of separations
            List<BufferedImage> separatedImages = pg.drawContents(params, colorants);
            
            try {
                int number = 0;
                for (BufferedImage image : separatedImages) {
	                File outputfile = new File(sOutput + number + "-" + inks.get(number).getColorantName() + ".png" );
	                ImageIO.write(image, "png", outputfile);
	                number++;
                }
            } catch (IOException e) {
            }
        }      
        finally {
            lib.delete();
        }
    }
}

File: DotNET/Sample_Source/Images/DrawSeparations/DrawSeparations.cs
using System;
using System.Collections.Generic;
using SkiaSharp;
using Datalogics.PDFL;
using System.IO;

/*
 * This sample demonstrates for drawing a list of grayscale separations from a PDF file.
 *
 * For more detail see the description of the DrawSeparations sample program on our Developer’s site, 
 * http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/net-core-sample-programs/manipulating-graphics-and-separating-colors-for-images
 * 
 * Copyright (c) 2007-2023, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */

namespace DrawSeparations
{
    class DrawSeparations
    {
        static void Main(string[] args)
        {

            Console.WriteLine("DrawSeparations Sample:");


            // ReSharper disable once UnusedVariable
            using (Library lib = new Library())
            {
                Console.WriteLine("Initialized the library.");

                String sInput = Library.ResourceDirectory + "Sample_Input/ducky.pdf";
                String sOutput = "DrawSeparations-out";

                if (args.Length > 0)
                    sInput = args[0];

                if (args.Length > 1)
                    sOutput = args[1];

                Console.WriteLine("Input file: " + sInput + ", writing output using prefix: " + sOutput);

                Document doc = new Document(sInput);
                Page pg = doc.GetPage(0);

                // Get all inks that are present on the page
                IList<Ink> inks = pg.ListInks();
                List<SeparationColorSpace> colorants = new List<SeparationColorSpace>();

                // Here we decide, which inks should be drawn
                foreach (Ink theInk in inks)
                {
                    // note: if the Ink can't be found in page's resources,
                    // default tintTransform and alternate will be used
                    colorants.Add(new SeparationColorSpace(pg, theInk));
                }

                double width = pg.MediaBox.Right - pg.MediaBox.Left;
                double height = pg.MediaBox.Top - pg.MediaBox.Bottom;

                // Must invert the page to get from PDF with origin at lower left,
                // to a bitmap with the origin at upper right.
                Matrix matrix = new Matrix().Scale(1, -1).Translate(-pg.MediaBox.Left, -pg.MediaBox.Top);

                DrawParams parms = new DrawParams();
                parms.Matrix = matrix;
                parms.DestRect = new Rect(0, 0, width, height);
                parms.Flags = DrawFlags.DoLazyErase | DrawFlags.UseAnnotFaces;

                double bottom = 0;
                if (pg.MediaBox.Bottom != 0)
                {
                    bottom = pg.MediaBox.Bottom < 0 ? -pg.MediaBox.Top : pg.MediaBox.Top;
                }

                parms.UpdateRect = new Rect(pg.MediaBox.Left, bottom, width, height);

                // Acquiring list of separations
                List<SKBitmap> separatedColorChannels = pg.DrawContents(parms, colorants);

                for (int i = 0; i < separatedColorChannels.Count; ++i)
                {
                    using (FileStream f = File.OpenWrite(sOutput + i + "-" + colorants[i].SeparationName + ".png"))
                        separatedColorChannels[i].Encode(SKEncodedImageFormat.Png, 100).SaveTo(f);
                }
            }
        }
    }
}
